api/server/router/container/container_routes.go

```diff
@@ -497,8 +497,8 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 			hostConfig.IpcMode = container.IpcMode("shareable")
 		}
 	}
-	if hostConfig != nil && versions.LessThan(version, "1.41") {
-		// Older clients expect the default to be "host"
+	if hostConfig != nil && versions.LessThan(version, "1.41") && !s.cgroup2 {
```

Comment:
I would just use `cgroups.IsCgroup2UnifiedMode()` right here. It's a very cheap call and we're definitely on the same host, and there's no need to pass the data around

Reply:
... and maybe put `!cgroups.IsCgroup2UnifiedMode()` before the version check since it's a cheaper check to make.

Reply:
I'm not sure. The router interface is designed to be decoupled from the host codes.

Reply:
You're right.

I think the best way then is to leave all this as is, just replacing `opts.daemon.RawSysInfo(true).CgroupUnified` with `cgroups.IfCgroupUnifiedMode()` in `initRouter()` as suggested below in https://github.com/moby/moby/pull/41072/commits/938a00ffcd50932a40c92e97d875274a12599709#r439165398.

There are a few other ways to handle it, such as forcibly set hostconfig.CgroupfsMode to private in container's hostconfig if it is not set and cgroupv2 is used, but I like it less.

Reply:
updated

Reply:
reverted due to compilation failure on Windows